File: ios-cuantrack/CuanTrack/Models/Transaction.swift
//
//  Transaction.swift
//  CuanTrack
//

import Foundation

enum TransactionType: String, Codable, CaseIterable {
    case expense = "Pengeluaran"
    case income = "Pemasukan"
    case transfer = "Transfer"
}

enum TransactionCategory: String, Codable, CaseIterable {
    case food = "Makanan"
    case transport = "Transport"
    case shopping = "Belanja"
    case bills = "Tagihan"
    case entertainment = "Hiburan"
    case other = "Lainnya"

    var icon: String {
        switch self {
        case .food: "fork.knife"
        case .transport: "car.fill"
        case .shopping: "bag.fill"
        case .bills: "doc.text.fill"
        case .entertainment: "tv.fill"
        case .other: "ellipsis.circle.fill"
        }
    }
}

struct Transaction: Identifiable, Codable {
    let id: UUID
    var type: TransactionType
    var category: TransactionCategory
    var amount: Double
    var accountName: String
    var date: Date
    var note: String
    var isRecurring: Bool

    init(
        id: UUID = UUID(),
        type: TransactionType = .expense,
        category: TransactionCategory = .other,
        amount: Double = 0,
        accountName: String = "",
        date: Date = Date(),
        note: String = "",
        isRecurring: Bool = false
    ) {
        self.id = id
        self.type = type
        self.category = category
        self.amount = amount
        self.accountName = accountName
        self.date = date
        self.note = note
        self.isRecurring = isRecurring
    }
}

File: android-cuantrack/app/src/main/java/com/rork/cuantrackandroid/models/Transaction.kt
package com.rork.cuantrackandroid.models

import androidx.compose.ui.graphics.Color
import java.time.Instant
import java.util.UUID

enum class TransactionType(val label: String) {
    EXPENSE("Pengeluaran"),
    INCOME("Pemasukan"),
    TRANSFER("Transfer");
}

enum class TransactionCategory(val label: String, val iconName: String) {
    FOOD("Makanan", "restaurant"),
    TRANSPORT("Transport", "directions_car"),
    SHOPPING("Belanja", "shopping_bag"),
    BILLS("Tagihan", "description"),
    ENTERTAINMENT("Hiburan", "tv"),
    OTHER("Lainnya", "more_horiz");

    val color: Color
        get() = when (this) {
            FOOD -> Color(0xFFF58738)
            TRANSPORT -> Color(0xFF3B99D9)
            SHOPPING -> Color(0xFFAB54D9)
            BILLS -> Color(0xFFDE5454)
            ENTERTAINMENT -> Color(0xFFF0338D)
            OTHER -> Color(0xFF64748B)
        }
}

data class Transaction(
    val id: UUID = UUID.randomUUID(),
    val type: TransactionType = TransactionType.EXPENSE,
    val category: TransactionCategory = TransactionCategory.OTHER,
    val amount: Double = 0.0,
    val accountName: String = "",
    val date: Instant = Instant.now(),
    val note: String = "",
    val isRecurring: Boolean = false
) {
    val formattedAmount: String
        get() {
            val nf = java.text.NumberFormat.getNumberInstance(java.util.Locale("id", "ID"))
            nf.maximumFractionDigits = 0
            return "Rp ${nf.format(amount)}"
        }

    val amountPrefix: String
        get() = if (type == TransactionType.INCOME) "+ " else "- "

    val amountColor: Color
        get() = if (type == TransactionType.INCOME) Color(0xFF16A34A) else Color(0xFFDC2626)
}
